File: 01_arrays_hashing/04_anagram_groups/solution.js
/*
Arrays & Hashing -> Anagram Groups
Neetcode 4 / 150
Leetcode 49

Javascript

Comments:
Strings are just arrays of characters. Once you sort them, anagrams are the same.

Time: O(m * nlog[n]) sort each word in the list
Space: O(m * n) sorted word for each word, map for all words at worst
*/

class Solution {
	/**
	 * @param {string[]} strs
	 * @return {string[][]}
	 */
	groupAnagrams(strs) {
		const groups = {};
		for (const word of strs) {
			const sorted = word.split("").sort().join();
			if (!groups[sorted]) {
				groups[sorted] = [];
			}
			groups[sorted].push(word);
		}
		return Object.values(groups);
	}
}

File: 01_arrays_hashing/04_anagram_groups/solution.py
'''
Arrays & Hashing -> Anagram Groups
Neetcode 4 / 150
Leetcode 49

Python

Comments:
This method is a bit more efficient and much easier to implement in python. 
Iterate through each word to count its letters. Can't do this with a dict because
you can't use mutable types as keys. ord(char) gives unicode value.

Time: O(m * n) loop through each word, each letter in each word
Space: O(m * n) array for each word, map for all words at worst
'''

from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        groups = defaultdict(list)
        for word in strs:
            count = [0] * 26
            for ltr in word:
                count[ord(ltr) - ord("a")] += 1
            groups[tuple(count)].append(word)
        return groups.values()